### PR TITLE
feat: add crd validation

### PR DESCRIPTION
**Pull request by thxCode**
**Fixes**:

https://github.com/rancher/harvester/issues/56

**Solution**:

- support to generate CRD validation spec.
- separate the CRD initialization from global setup.
- adjust the CRD initialization logic so that it doesn't overwrite the existing CRD.

### CODE REVIEW

File: pkg/apis/harvester.cattle.io/v1alpha1/template.go [hunk 1]
```diff
@@ -33,25 +49,27 @@ type VirtualMachineTemplateVersion struct {
 	Status VirtualMachineTemplateVersionStatus `json:"status,omitempty"`
 }
 
-type VirtualMachineTemplateSpec struct {
-	Description      string `json:"description,omitempty"`
-	DefaultVersionID string `json:"defaultVersionId,omitempty"`
-}
+type VirtualMachineTemplateVersionSpec struct {
```
Comment (guangbochen) [inline pkg/apis/harvester.cattle.io/v1alpha1/template.go:63]: would prefer to add shortName as `vmtemplateversion` and print its templateId and status.version.

**Review by guangbochen — CHANGES_REQUESTED** [review]

File: pkg/apis/harvester.cattle.io/v1alpha1/image.go [hunk 2]
```diff
@@ -22,18 +22,34 @@ type VirtualMachineImage struct {
 }
 
 type VirtualMachineImageSpec struct {
-	DisplayName string `json:"displayName"`
-	Description string `json:"description"`
-	URL         string `json:"url"`
-	SecretRef   string `json:"secretRef"`
+	// +kubebuilder:validation:Required
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/apis/harvester.cattle.io/v1alpha1/image.go:31]: Updated

File: pkg/apis/harvester.cattle.io/v1alpha1/template.go [hunk 3]
```diff
@@ -21,6 +21,22 @@ type VirtualMachineTemplate struct {
 	Status VirtualMachineTemplateStatus `json:"status,omitempty"`
 }
 
+type VirtualMachineTemplateSpec struct {
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/apis/harvester.cattle.io/v1alpha1/template.go:30]: Updated

File: pkg/apis/harvester.cattle.io/v1alpha1/template.go (hunk 1 above)
Comment (thxCode) [reply pkg/apis/harvester.cattle.io/v1alpha1/template.go:63]: Updated

File: Dockerfile.dapper [hunk 4]
```diff
@@ -24,13 +24,17 @@ RUN apt-get update -qq && apt-get install -y --no-install-recommends \
...
+        GO111MODULE=on go get sigs.k8s.io/controller-tools/cmd/controller-gen@v0.3.0; \
+    fi
 # -- for make rules
 
 # -- for dapper
 ENV GO111MODULE off
 ENV DAPPER_ENV REPO TAG DRONE_TAG CROSS
 ENV DAPPER_SOURCE /go/src/github.com/rancher/harvester/
-ENV DAPPER_OUTPUT ./bin ./dist
+ENV DAPPER_OUTPUT ./bin ./dist ./deploy
```
Comment (orangedeng) [inline Dockerfile.dapper]: Can we move `./deploy` to `./dist/deploy`.

Comment (orangedeng) [reply Dockerfile.dapper]: Why do we need to set `./deploy` output here? I see the crd manifests under this dir in this PR.

Comment (thxCode) [reply Dockerfile.dapper]: `./bin` and `./dist` are used to put some generated assets during the CI phase, like `test` report or other intermediate artifacts. However, these generated CRD manifests are not only some artifacts but also the pre-install resources of chart, so I choose to put them under `./deploy/charts/harvester/crds` dir. 

~Defined the dapper output here is to make sure we can execute `make generate-manifest`, should not consider this?~

Updated

**Review by orangedeng — APPROVED** [review]
LGTM

**Review by guangbochen — APPROVED** [review]
LGTM

**Pull request merged**